core/common/src/test/java/alluxio/concurrent/CountLatchTest.java

```diff
@@ -0,0 +1,215 @@
+
+    inc.join();
+    Assert.assertTrue(inc.getBlockedTimeMillis() >= SLEEP_MILLIS);
+    Assert.assertEquals(1, mLatch.getState());
+  }
+
+  /**
+   * Tests that multiple blocked inc can be unblocked when unblockInc returns.
+   */
+  @Test
+  public void unblockMultipleInc() throws Exception {
+    Assert.assertEquals(0, mLatch.getState());
+    mLatch.awaitAndBlockInc();
+    Assert.assertEquals(-1, mLatch.getState());
+
+    List<BlockingThread> incThreads = new ArrayList<>();
+    final int numThreads = 10;
+    for (int i = 0; i < numThreads; i++) {
+      incThreads.add(new BlockingThread(() -> {
+        try {
+          mLatch.inc();
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      }));
+    }
+    for (BlockingThread t : incThreads) {
+      t.start();
+    }
+
+    Thread.sleep(SLEEP_MILLIS);
+    mLatch.unblockInc();
+
+    for (BlockingThread t : incThreads) {
+      t.join();
+      Assert.assertTrue(t.getBlockedTimeMillis() >= SLEEP_MILLIS);
+    }
+    Assert.assertEquals(numThreads, mLatch.getState());
+  }
+
+  /**
+   * Tests that awaitAndBlockInc is blocked when state is not 0.
+   */
+  @Test
+  public void await() throws Exception {
+    Assert.assertEquals(0, mLatch.getState());
+    mLatch.inc();
+    Assert.assertEquals(1, mLatch.getState());
+    mLatch.inc();
+    Assert.assertEquals(2, mLatch.getState());
+
+    BlockingThread await = new BlockingThread(mLatch::awaitAndBlockInc);
+    await.start();
+
+    Assert.assertEquals(STILL_BLOCKED, await.getBlockedTimeMillis());
+
+    Thread.sleep(SLEEP_MILLIS);
+    mLatch.dec();
+    Assert.assertEquals(1, mLatch.getState());
+    Assert.assertEquals(STILL_BLOCKED, await.getBlockedTimeMillis());
+
+    Thread.sleep(SLEEP_MILLIS);
+    mLatch.dec();
+    Assert.assertEquals(0, mLatch.getState());
+
+    await.join();
+    Assert.assertTrue(await.getBlockedTimeMillis() >= 2 * SLEEP_MILLIS);
+    Assert.assertEquals(-1, mLatch.getState());
+  }
+
+  /**
+   * Tests that calling dec without a paired inc will throw exception.
+   */
+  @Test
+  public void decWithoutInc() throws Exception {
+    mLatch.inc();
+    mLatch.dec();
+    mExpectedException.expect(Error.class);
+    mLatch.dec();
+  }
+
+  /**
+   * Tests that unblock without a paired awaitAndBlockInc will throw exception.
+   */
+  @Test
+  public void unblockWithoutAwait() {
+    mLatch.awaitAndBlockInc();
+    mLatch.unblockInc();
+    mExpectedException.expect(Error.class);
+    mLatch.unblockInc();
+  }
+
+  /**
+   * Tests that inc can be interrupted.
+   */
+  @Test
+  public void interruptInc() throws Exception {
+    mLatch.awaitAndBlockInc();
+    Thread inc = new Thread(() -> {
+      try {
+        mLatch.inc();
+      } catch (InterruptedException e) {
+        // Expected.
+      }
+    });
+    inc.start();
+    inc.join(SLEEP_MILLIS);
+    Assert.assertEquals(Thread.State.WAITING, inc.getState());
+    inc.interrupt();
+    inc.join();
+  }
+}
```

Comment:
Could you add a test to validate calling awaitAndBlockInc() does not cause, following inc()s to block.

Reply:
thanks for catching the missing test case